### PR TITLE
NIFI-981 [REVIEW-ONLY]: Initial addition of HiveQL processors

### PR DESCRIPTION
**Pull request by mattyb149**
I'd like to get some feedback on the ExecuteHiveQL and PutHiveQL processors before I start adding Kerberos support and such. I will incorporate any comments into the code and then re-issue a PR when the processors are ready for final review.

### CODE REVIEW

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+/**
+ * Implementation for Database Connection Pooling Service used for Apache Hive connections. Apache DBCP is used for connection pooling functionality.
+ */
+@Tags({"hive", "dbcp", "jdbc", "database", "connection", "pooling", "store"})
+@CapabilityDescription("Provides Database Connection Pooling Service for Apache Hive. Connections can be asked from pool and returned after usage.")
+public class HiveConnectionPool extends AbstractControllerService implements HiveDBCPService {
+
+    public static final PropertyDescriptor DATABASE_URL = new PropertyDescriptor.Builder()
+            .name("Database Connection URL")
```
Comment (olegz) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java]: Matt, one thing that has been actually discussed in the couple of recent PRs was to start embracing ```.displayName(string)```. The main reason is that we are currently coupling what user's see to the actual field names. The problem with this approach is obvious. If at some point down the road we realize that we should have named something differently we can't do that since all the templates that are using it will break.

With ```displayName``` we would essentially decouple what user sees from the the actual property name (even if initially they are the same). This way we are free to modify it in the future every way we want to make it more user friendly without breaking backwards compatibility.

Obviously there is no precedence for this other then me and @alopresto joining forces in advocating it. So your choice, but I figured I'll mention. ;)

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+        List<PropertyDescriptor> props = new ArrayList<>();
+        props.add(DATABASE_URL);
+        props.add(HIVE_CONFIGURATION_RESOURCES);
+        props.add(DB_USER);
+        props.add(DB_PASSWORD);
+        props.add(MAX_WAIT_TIME);
+        props.add(MAX_TOTAL_CONNECTIONS);
+        properties = Collections.unmodifiableList(props);
+    }
```
Comment (olegz) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java]: Given what we know and that is this method will get executed several times before the processor will actually shows up on the canvas I'd suggest using a bit different approach. For example: https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/PublishAMQP.java#L102

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+            .required(true)
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .sensitive(false)
+            .build();
+
+    public static final PropertyDescriptor MAX_TOTAL_CONNECTIONS = new PropertyDescriptor.Builder()
+            .name("Max Total Connections")
+            .description("The maximum number of active connections that can be allocated from this pool at the same time, "
+                    + " or negative for no limit.")
+            .defaultValue("8")
```
Comment (olegz) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java]: Just curious, why 8? I mean as a default I still think I'd chose 1 (safest option) and have user bump it up as needed

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+        final String dburl = context.getProperty(DATABASE_URL).getValue();
+
+        dataSource.setMaxWait(maxWaitMillis);
+        dataSource.setMaxActive(maxTotal);
+
+        dataSource.setUrl(dburl);
+        dataSource.setUsername(user);
+        dataSource.setPassword(passw);
+    }
```
Comment (olegz) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java]: Hmm, since it appears you are using DBCP, shouldn't you use the DBCP ControllerService?

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+    /**
+     * Shutdown pool, close all open connections.
+     */
+    @OnDisabled
+    public void shutdown() {
+        try {
+            dataSource.close();
+        } catch (final SQLException e) {
+            throw new ProcessException(e);
```
Comment (olegz) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java]: Do we really want to do that? What's the recovery from that? I think just logging it with WARN would be enough. It seems like a common pattern

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/processors/hive/ExecuteHiveQL.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        final Set<Relationship> r = new HashSet<>();
+        r.add(REL_SUCCESS);
+        r.add(REL_FAILURE);
+        relationships = Collections.unmodifiableSet(r);
+
+        final List<PropertyDescriptor> pds = new ArrayList<>();
+        pds.add(HIVE_DBCP_SERVICE);
+        pds.add(HIVEQL_SELECT_QUERY);
+        propDescriptors = Collections.unmodifiableList(pds);
+    }
```
Comment (olegz) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/processors/hive/ExecuteHiveQL.java]: This one will be executed multiple times as well. Lot of unnecessary work for GC for no reason.

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/processors/hive/PutHiveQL.java [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+        _propertyDescriptors.add(BATCH_SIZE);
+        _propertyDescriptors.add(CHARSET);
+        propertyDescriptors = Collections.unmodifiableList(_propertyDescriptors);
+
+        Set<Relationship> _relationships = new HashSet<>();
+        _relationships.add(REL_SUCCESS);
+        _relationships.add(REL_FAILURE);
+        _relationships.add(REL_RETRY);
+        relationships = Collections.unmodifiableSet(_relationships);
+    }
```
Comment (olegz) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/processors/hive/PutHiveQL.java]: This looks familiar. Consider making it cocnsistent across

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/processors/hive/PutHiveQL.java [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+                    case Types.CHAR:
+                    case Types.VARCHAR:
+                    case Types.LONGNVARCHAR:
+                    case Types.LONGVARCHAR:
+                        stmt.setString(parameterIndex, parameterValue);
+                        break;
+                    default:
+                        stmt.setObject(parameterIndex, parameterValue, jdbcType);
+                        break;
+                }
```
Comment (olegz) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/processors/hive/PutHiveQL.java]: As much as I don't like the above, we know now that setObject does not always do what it was intended to do so we stuck with it. . . Argh. . . just venting ;)

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java (hunk 3 above)
Comment (mattyb149) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java]: Copied this from DBCPConnectionPool, had no reason to change it

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java (hunk 5 above)
Comment (mattyb149) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java]: Not sure why/if this needs to be done, another copy-paste from DBCPConnectionPool

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+                    + " for a connection to be returned before failing, or -1 to wait indefinitely. ")
+            .defaultValue("500 millis")
+            .required(true)
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .sensitive(false)
+            .build();
+
+    public static final PropertyDescriptor MAX_TOTAL_CONNECTIONS = new PropertyDescriptor.Builder()
+            .name("Max Total Connections")
+            .description("The maximum number of active connections that can be allocated from this pool at the same time, "
```
Comment (markap14) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java]: This seems odd to me - why is this required with a negative value indicating no limit? Why not just make it optional and use a POSITIVE_INTEGER_VALIDATOR? Then, if you want no limit, just don't enter a value.

Comment (markap14) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java]: Though I guess we are setting a Default Value here, so if you remove the value, it will just reset to 8. So I think I understand the reason for allowing a negative value now. We should probably consider allowing an optional property to have its value removed, even if there is a default value.

Comment (mattyb149) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java]: This is another copy-paste from DBCPConnectionPool, I imagine it's so the value can be put immediately into the DBCP DataSource. I agree it's unnecessarily awkward, will make the changes you suggest. This would make it inconsistent with the DBCPConnectionPool, but I think that should be ok

Comment (olegz) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/dbcp/hive/HiveConnectionPool.java]: Personally I'd use 1 as a default and use 0 (a very common and well known convention to depict "no limit").
